pkg/core/alias.go

```diff
@@ -22,24 +22,47 @@ var (
 	SetLogger             = kube_log.SetLogger
 	Now                   = time.Now
 
-	SetupSignalHandler = func() (context.Context, context.Context) {
+	SetupSignalHandler = func() (context.Context, context.Context, <-chan struct{}) {
 		gracefulCtx, gracefulCancel := context.WithCancel(context.Background())
 		ctx, cancel := context.WithCancel(context.Background())
 		c := make(chan os.Signal, 3)
-		signal.Notify(c, syscall.SIGINT, syscall.SIGTERM)
+		signal.Notify(c, syscall.SIGINT, syscall.SIGTERM, syscall.SIGUSR2)
+		usr2Notify := make(chan struct{}, 1)
 		go func() {
 			logger := Log.WithName("runtime")
-			s := <-c
-			logger.Info("received signal, stopping instance gracefully", "signal", s.String())
-			gracefulCancel()
-			s = <-c
-			logger.Info("received second signal, stopping instance", "signal", s.String())
-			cancel()
-			s = <-c
-			logger.Info("received third signal, force exit", "signal", s.String())
-			os.Exit(1)
+
+			var stopSignalCancel func(os.Signal)
+			thirdStopSignal := func(signal os.Signal) {
+				logger.Info("received third signal, force exit", "signal", signal.String())
+				os.Exit(1)
+			}
+			secondStopSignal := func(signal os.Signal) {
+				logger.Info("received second signal, stopping instance", "signal", signal.String())
+				cancel()
+				stopSignalCancel = thirdStopSignal
+			}
+			firstStopSignal := func(signal os.Signal) {
+				logger.Info("received signal, stopping instance gracefully", "signal", signal.String())
+				gracefulCancel()
+				close(usr2Notify)
+				usr2Notify = nil
+				stopSignalCancel = secondStopSignal
+			}
+			stopSignalCancel = firstStopSignal
+			for {
+				s := <-c
+				switch s {
+				case syscall.SIGINT, syscall.SIGTERM:
+					stopSignalCancel(s)
+				case syscall.SIGUSR2:
+					select {
+					case usr2Notify <- struct{}{}:
+					default:
+					}
+				}
+			}
```

Comment:
How about something that doesn't iterate over function references?

```go
	signalReceived := 0
	for {
		s:=<-c
		switch s {
		case syscall.SIGINT, syscall.SIGTERM:
			signalReceived += 1
			logger.Info("received signal, stopping instance gracefully", "signal", s.String(), "signalCount", signalReceived)
		case syscall.SIGUSR2:
			select {
			case usr2Notify <- struct{}{}:
			default:
			}
			break
		}
		switch signalReceived {
		case 1:
			gracefulCancel()
			close(usr2Notify)
			usr2Notify = nil
		case 2:
			cancel()
		case 3:
			s.Exit(1)
		default:
			// ??
		}
	}
```

Reply:
I think having a direct pointer from signal -> transition makes it easier to create the state machine in my head but if you both find this way easier I can change it

Reply:
Yeah I find your new way much easier to read